File: api/routes/tax_forms.py
"""
Tax-forms routes:
  GET  /tax-forms/compute             — instant JSON review (no PDF generation)
  POST /reports/tax-forms             — start ZIP-generation background job
  GET  /reports/tax-forms/{job_id}    — poll job status
  GET  /reports/tax-forms/{job_id}/download — download the ZIP
"""

import sys
import threading
import uuid
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import FileResponse, Response

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "scripts"))

from api.auth import get_current_user_optional
from api.db import get_filing_details, save_filing_details

router = APIRouter(tags=["tax-forms"])

_jobs: dict[str, dict] = {}


@router.get("/filing-details")
def get_filing_details_route(
    tax_year: int = Query(default=2025),
    current_user: dict | None = Depends(get_current_user_optional),
):
    if not current_user:
        raise HTTPException(401, "Authentication required")
    return get_filing_details(current_user["id"], tax_year)


@router.put("/filing-details")
def save_filing_details_route(
    payload: dict,
    tax_year: int = Query(default=2025),
    current_user: dict | None = Depends(get_current_user_optional),
):
    if not current_user:
        raise HTTPException(401, "Authentication required")
    save_filing_details(current_user["id"], tax_year, payload)
    return {"ok": True}


@router.get("/tax-forms/preview-pdf")
def preview_tax_form_pdf(
    tax_year: int = Query(default=2025),
    current_user: dict | None = Depends(get_current_user_optional),
):
    """Return a filled Form 1040 PDF directly. Uses cached IRS base form when available."""
    if not current_user:
        raise HTTPException(401, "Authentication required")
    try:
        from generate_tax_forms import generate_preview_pdf  # noqa: E402
        pdf_bytes = generate_preview_pdf(current_user["id"], tax_year)
    except ValueError as exc:
        raise HTTPException(503, str(exc))
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": "inline; filename=form_1040_preview.pdf"},
    )


@router.get("/tax-forms/compute")
def compute_tax_forms(
    tax_year: int = Query(default=2025),
    current_user: dict | None = Depends(get_current_user_optional),
):
    """Return computed tax figures as JSON — no PDF generation, runs instantly."""
    if not current_user:
        raise HTTPException(401, "Authentication required")
    from generate_tax_forms import compute_tax_figures  # noqa: E402
    return compute_tax_figures(current_user["id"], tax_year)


def _run_job(job_id: str, user_id: str, tax_year: int) -> None:
    try:
        from generate_tax_forms import FormPackageGenerator  # noqa: E402

        messages: list[str] = []

        def _progress(msg: str):
            messages.append(msg)
            _jobs[job_id]["progress"] = msg

        gen  = FormPackageGenerator(user_id, tax_year)
        path = gen.generate(progress_cb=_progress)
        _jobs[job_id].update({
            "status":   "complete",
            "zip_path": path,
            "zip_name": Path(path).name,
            "messages": messages,
            "error":    None,
        })
    except Exception as exc:
        _jobs[job_id].update({
            "status":   "error",
            "zip_path": None,
            "zip_name": None,
            "error":    str(exc),
        })


@router.post("/reports/tax-forms")
def start_tax_forms(
    tax_year: int = Query(default=2025),
    current_user: dict | None = Depends(get_current_user_optional),
):
    if not current_user:
        raise HTTPException(401, "Authentication required to generate tax forms")
    job_id = str(uuid.uuid4())
    _jobs[job_id] = {
        "status":   "running",
        "progress": "Starting…",
        "zip_path": None,
        "zip_name": None,
        "error":    None,
    }
    threading.Thread(
        target=_run_job, args=(job_id, current_user["id"], tax_year), daemon=True
    ).start()
    return {"job_id": job_id}


@router.get("/reports/tax-forms/{job_id}")
def get_tax_forms_status(job_id: str):
    job = _jobs.get(job_id)
    if job is None:
        raise HTTPException(404, f"Job '{job_id}' not found")
    # Return everything except the raw path for security
    return {
        "status":   job["status"],
        "progress": job.get("progress"),
        "zip_name": job.get("zip_name"),
        "error":    job.get("error"),
    }


@router.get("/reports/tax-forms/{job_id}/download")
def download_tax_forms(job_id: str):
    """
    No auth header needed — the UUID job_id is the bearer of authority.
    Called via a plain browser <a href> so Authorization header is unavailable.
    """
    job = _jobs.get(job_id)
    if job is None:
        raise HTTPException(404, "Job not found")
    if job["status"] != "complete":
        raise HTTPException(400, f"Job status is '{job['status']}' — not ready for download")
    zip_path = Path(job["zip_path"])
    if not zip_path.exists():
        raise HTTPException(500, "Package file missing from disk")
    return FileResponse(
        path=str(zip_path),
        media_type="application/zip",
        filename=zip_path.name,
    )

File: backend-ts/src/routes/taxForms.ts
import fs from "node:fs";
import path from "node:path";
import crypto from "node:crypto";
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { AppError } from "../lib/errors";
import { projectPaths } from "../lib/paths";
import { getFilingDetails, saveFilingDetails, type FilingDetails } from "../db/filingDetailsRepo";
import { computeTaxFigures, buildFormPackage, loadAllUserData } from "../domain/taxForms/index";
import { fillIrsForms, fillSingleIrsForm } from "../domain/taxForms/fillIrsForms";

type JobState = {
  status: "running" | "complete" | "error";
  progress: string;
  zip_path: string | null;
  zip_name: string | null;
  error: string | null;
};

const jobs = new Map<string, JobState>();

const TaxYearQuerySchema = z.object({
  tax_year: z.coerce.number().int().optional()
});

const PreviewQuerySchema = z.object({
  tax_year: z.coerce.number().int().optional(),
  form: z.string().regex(/^[A-Za-z0-9_]+$/).optional()
});

const JobParamsSchema = z.object({
  jobId: z.string().min(1)
});

const FilingDetailsBodySchema = z.object({
  pec_fund_taxpayer: z.boolean().optional(),
  pec_fund_spouse: z.boolean().optional(),
  direct_deposit_routing: z.union([z.string(), z.null()]).optional(),
  direct_deposit_account: z.union([z.string(), z.null()]).optional(),
  direct_deposit_type: z.union([z.string(), z.null()]).optional(),
  allow_third_party: z.boolean().optional(),
  designee_name: z.union([z.string(), z.null()]).optional(),
  designee_phone: z.union([z.string(), z.null()]).optional(),
  designee_pin: z.union([z.string(), z.null()]).optional()
}).strict();

async function buildPreviewPdfBytes(userId: string, taxYear: number): Promise<Uint8Array> {
  const data = loadAllUserData(userId, taxYear);
  const figures = computeTaxFigures(userId, taxYear);
  return fillIrsForms(figures.computed, figures.display_name, taxYear, data);
}

async function runJob(jobId: string, userId: string, taxYear: number): Promise<void> {
  try {
    fs.mkdirSync(path.join(projectPaths.root, "state", "tax_form_packages"), { recursive: true });
    const zipName = `tax_forms_${taxYear}_${jobId.slice(0, 8)}.zip`;
    const zipPath = path.join(projectPaths.root, "state", "tax_form_packages", zipName);
    const zipBytes = await buildFormPackage(userId, taxYear);
    fs.writeFileSync(zipPath, zipBytes);
    jobs.set(jobId, {
      status: "complete",
      progress: "Package generated",
      zip_path: zipPath,
      zip_name: zipName,
      error: null,
    });
  } catch (error) {
    jobs.set(jobId, {
      status: "error",
      progress: "Package generation failed",
      zip_path: null,
      zip_name: null,
      error: (error as Error).message,
    });
  }
}

export async function registerTaxFormsRoutes(app: FastifyInstance): Promise<void> {
  app.get("/filing-details", { preHandler: app.authenticate }, async (request) => {
    const query = TaxYearQuerySchema.parse(request.query ?? {});
    const taxYear = query.tax_year ?? 2025;
    const userId = request.currentUser?.id;
    if (!userId) {
      throw new AppError(401, "Authentication required");
    }
    return getFilingDetails(userId, taxYear);
  });

  app.put("/filing-details", { preHandler: app.authenticate }, async (request) => {
    const query = TaxYearQuerySchema.parse(request.query ?? {});
    const taxYear = query.tax_year ?? 2025;
    const userId = request.currentUser?.id;
    if (!userId) {
      throw new AppError(401, "Authentication required");
    }

    const payload = FilingDetailsBodySchema.parse(request.body ?? {}) as FilingDetails;
    saveFilingDetails(userId, taxYear, payload);
    return { ok: true };
  });

  app.get("/tax-forms/preview-pdf", { preHandler: app.authenticate }, async (request, reply) => {
    const query = PreviewQuerySchema.parse(request.query ?? {});
    const taxYear = query.tax_year ?? 2025;
    const userId = request.currentUser?.id;
    if (!userId) throw new AppError(401, "Authentication required");

    if (query.form) {
      const data = loadAllUserData(userId, taxYear);
      const figures = computeTaxFigures(userId, taxYear);
      const pdfBytes = await fillSingleIrsForm(query.form, figures.computed, figures.display_name, taxYear, data);
      const filename = `${query.form}_${taxYear}.pdf`;
      return reply
        .type("application/pdf")
        .header("Content-Disposition", `inline; filename=${filename}`)
        .send(Buffer.from(pdfBytes));
    }

    const pdfBytes = await buildPreviewPdfBytes(userId, taxYear);
    return reply
      .type("application/pdf")
      .header("Content-Disposition", "inline; filename=form_1040_preview.pdf")
      .send(Buffer.from(pdfBytes));
  });

  app.get("/tax-forms/compute", { preHandler: app.authenticate }, async (request) => {
    const query = TaxYearQuerySchema.parse(request.query ?? {});
    const taxYear = query.tax_year ?? 2025;
    const userId = request.currentUser?.id;
    if (!userId) {
      throw new AppError(401, "Authentication required");
    }

    const figures = computeTaxFigures(userId, taxYear);
    return {
      ...figures,
      filing_details: getFilingDetails(userId, taxYear),
      summary: { counts: {} },
    };
  });

  app.post("/reports/tax-forms", { preHandler: app.authenticate }, async (request) => {
    const query = TaxYearQuerySchema.parse(request.query ?? {});
    const taxYear = query.tax_year ?? 2025;
    const userId = request.currentUser?.id;
    if (!userId) {
      throw new AppError(401, "Authentication required to generate tax forms");
    }

    const jobId = crypto.randomUUID();
    jobs.set(jobId, {
      status: "running",
      progress: "Starting...",
      zip_path: null,
      zip_name: null,
      error: null
    });

    queueMicrotask(() => void runJob(jobId, userId, taxYear));

    return { job_id: jobId };
  });

  app.get("/reports/tax-forms/:jobId", async (request) => {
    const { jobId } = JobParamsSchema.parse(request.params ?? {});
    const job = jobs.get(jobId);
    if (!job) {
      throw new AppError(404, `Job '${jobId}' not found`);
    }

    return {
      status: job.status,
      progress: job.progress,
      zip_name: job.zip_name,
      error: job.error
    };
  });

  app.get("/reports/tax-forms/:jobId/download", async (request, reply) => {
    const { jobId } = JobParamsSchema.parse(request.params ?? {});
    const job = jobs.get(jobId);
    if (!job) {
      throw new AppError(404, "Job not found");
    }
    if (job.status !== "complete") {
      throw new AppError(400, `Job status is '${job.status}' — not ready for download`);
    }
    if (!job.zip_path || !fs.existsSync(job.zip_path)) {
      throw new AppError(500, "Package file missing from disk");
    }

    const content = fs.readFileSync(job.zip_path);
    return reply.type("application/zip").header("Content-Disposition", `attachment; filename=${job.zip_name}`).send(content);
  });
}
